pkg/querier/queryrange/roundtrip.go

```diff
@@ -72,60 +82,75 @@ func MergeMiddlewares(middleware ...Middleware) Middleware {
 	})
 }
 
+// NewTripperware returns a Tripperware configured with middlewares to limit, align, split, retry and cache requests.
+func NewTripperware(cfg Config, log log.Logger, limits Limits, codec Codec, cacheExtractor Extractor) (frontend.Tripperware, error) {
+	queryRangeMiddleware := []Middleware{LimitsMiddleware(limits)}
+	if cfg.AlignQueriesWithStep {
+		queryRangeMiddleware = append(queryRangeMiddleware, InstrumentMiddleware("step_align"), StepAlignMiddleware)
+	}
+	if cfg.SplitQueriesByDay {
+		queryRangeMiddleware = append(queryRangeMiddleware, InstrumentMiddleware("split_by_day"), SplitByDayMiddleware(limits, codec))
+	}
+	if cfg.CacheResults {
+		queryCacheMiddleware, err := NewResultsCacheMiddleware(log, cfg.ResultsCacheConfig, limits, codec, cacheExtractor)
+		if err != nil {
+			return nil, err
+		}
+		queryRangeMiddleware = append(queryRangeMiddleware, InstrumentMiddleware("results_cache"), queryCacheMiddleware)
+	}
+	if cfg.MaxRetries > 0 {
+		queryRangeMiddleware = append(queryRangeMiddleware, InstrumentMiddleware("retry"), NewRetryMiddleware(log, cfg.MaxRetries))
+	}
+
+	return frontend.Tripperware(func(next http.RoundTripper) http.RoundTripper {
+		// Finally, if the user selected any query range middleware, stitch it in.
+		if len(queryRangeMiddleware) > 0 {
+			return NewRoundTripper(next, codec, queryRangeMiddleware...)
+		}
+		return next
+	}), nil
+}
+
 type roundTripper struct {
 	next    http.RoundTripper
 	handler Handler
-	limits  Limits
+	codec   Codec
 }
 
-// NewRoundTripper wraps a QueryRange Handler and allows it to send requests
-// to a http.Roundtripper.
-func NewRoundTripper(next http.RoundTripper, handler Handler, limits Limits) http.RoundTripper {
-	return roundTripper{
-		next:    next,
-		handler: handler,
-		limits:  limits,
+// NewRoundTripper merges a set of middlewares into an handler, then inject it into the `next` roundtripper
+// using the codec to translate requests and responses.
+func NewRoundTripper(next http.RoundTripper, codec Codec, middlewares ...Middleware) http.RoundTripper {
+	transport := roundTripper{
+		next:  next,
+		codec: codec,
 	}
+	transport.handler = MergeMiddlewares(middlewares...).Wrap(&transport)
+	return transport
 }
 
 func (q roundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
 	if !strings.HasSuffix(r.URL.Path, "/query_range") {
 		return q.next.RoundTrip(r)
 	}
 
-	request, err := parseRequest(r)
+	request, err := q.codec.DecodeRequest(r.Context(), r)
 	if err != nil {
 		return nil, err
 	}
-	request.LogToSpan(r.Context())
 
-	userid, err := user.ExtractOrgID(r.Context())
-	if err != nil {
-		return nil, err
-	}
-
-	maxQueryLen := q.limits.MaxQueryLength(userid)
-	queryLen := timestamp.Time(request.End).Sub(timestamp.Time(request.Start))
-	if maxQueryLen != 0 && queryLen > maxQueryLen {
-		return nil, httpgrpc.Errorf(http.StatusBadRequest, validation.ErrQueryTooLong, queryLen, maxQueryLen)
-	}
+	LogToSpan(r.Context(), request)
```

Comment:
FYI @jtlisi @gouthamve  I had to rebased to master and re-introduce this.

see https://github.com/cortexproject/cortex/pull/1734/files#diff-19934b07a15ed5212e1637c543973bd1R79